File: solutions/python/protein-translation/protein_translation.py
def proteins(strand):
    """Translate RNA sequences into proteins.

    :param strand: string - the RNA strand
    :return: [string] - a list of proteins
    """

    STOP_CODON = "STOP"

    def from_codon(codon):
        match codon:
            case c if c in ["AUG"]:
                return "Methionine"
            case c if c in ["UUU", "UUC"]:
                return "Phenylalanine"
            case c if c in ["UUA", "UUG"]:
                return "Leucine"
            case c if c in ["UCU", "UCC", "UCA", "UCG"]:
                return "Serine"
            case c if c in ["UAU", "UAC"]:
                return "Tyrosine"
            case c if c in ["UGU", "UGC"]:
                return "Cysteine"
            case c if c in ["UGG"]:
                return "Tryptophan"
            case c if c in ["UAA", "UAG", "UGA"]:
                return STOP_CODON

    result = []
    for i in range(0, len(strand), 3):
        amino_acid = from_codon(strand[i : i + 3])
        if amino_acid == STOP_CODON:
            break
        result.append(amino_acid)
    return result

File: solutions/javascript/protein-translation/protein-translation.js
/**
 * Translate RNA sequences into proteins.
 *
 * @param {string} strand
 * @returns {string[]}
 */
export const translate = (strand) => {
  const result = [];
  if (!strand) return result;
  for (let chunk of strand.match(/.{1,3}/g)) {
    let aminoAcid = MAPPING[chunk];
    if (!aminoAcid) throw new Error("Invalid codon");
    if (aminoAcid === STOP_CODON) break;
    result.push(aminoAcid);
  }
  return result;
};
const STOP_CODON = "STOP";
/**
 * @type {Record<string,string}
 */
const MAPPING = {
  UGU: "Cysteine",
  UGC: "Cysteine",
  UUA: "Leucine",
  UUG: "Leucine",
  AUG: "Methionine",
  UUU: "Phenylalanine",
  UUC: "Phenylalanine",
  UCU: "Serine",
  UCC: "Serine",
  UCA: "Serine",
  UCG: "Serine",
  UGG: "Tryptophan",
  UAU: "Tyrosine",
  UAC: "Tyrosine",
  UAA: STOP_CODON,
  UAG: STOP_CODON,
  UGA: STOP_CODON,
};
